src/distPlus/distribution/Discrete.re

```diff
@@ -87,68 +100,80 @@ let combineAlgebraically =
 
   let combinedShape = XYShape.T.fromZippedArray(rxys);
 
-  make(combinedShape, combinedIntegralSum);
+  make(combinedShape, combinedIntegralSum, None);
 };
 
-let mapY = (~knownIntegralSumFn=previousKnownIntegralSum => None, fn, t: t) => {
-  let u = E.O.bind(_, knownIntegralSumFn);
+let mapY = (~integralSumCacheFn=_ => None,
+            ~integralCacheFn=_ => None,
+            fn, t: t) => {
   let yMapFn = shapeMap(XYShape.T.mapY(fn));
 
-  t |> yMapFn |> updateKnownIntegralSum(u(t.knownIntegralSum));
+  t
+  |> yMapFn
+  |> updateIntegralSumCache(E.O.bind(t.integralSumCache, integralSumCacheFn))
+  |> updateIntegralCache(E.O.bind(t.integralCache, integralCacheFn));
```

Comment:
Order is really important here. I think for a few intuitive reasons I'd prefer these lines not be a chain line this, it seems weird to me to do is this way. Different from,
```
let newXyShape = t |> shapeMap(XYShape.T.mapY(fn));
let newIntegralSumCache = integralSumCache |> E.O.bind(_, integralSumCacheFn)
...
make(~xyShape=newXyShape, ~integralSumCache=Some(newIntegralSumCache)...)
```